Support mock.patch class decorator

ukinau opened the pull request:
Change so that parameterised.expand check function is decorated by mock.patch or not by the function object's property (patching). 
If the function is decorated by mock.patch, it try to re-decorate the function by mock.patch after decorate it with dummy decoration so as mock.patch to decorate the function be generated by parameterised.expand as a all different function, and then make original patching list empty not to be evaluated.
By doing this approach, this PR fixed issue #52 

If owner of parameterised module think it's reasonable to support mock.patch class decorator, I will write unit test for this PR.

Fixes #52

wolever commented:
Hey! Thanks for the PR! This is definitely a thing parameterized should support.

I've left a couple comments on the implementation, but mostly looks good. Please include some tests and I'll get this merged!

wolever commented on parameterized/parameterized.py:
```diff
@@ -40,6 +40,31 @@ def make_method(func, instance, type):
+        func = dummy_wrapper(func)
+        tmp_patchings = func.patchings
+        delattr(func, 'patchings')
+        for i, patch_obj in enumerate(tmp_patchings):
```
The `tmp_patchings` is redundant; can just use `enumerate(func.patchings)`

ukinau replied:
Here, actually we can't just use `enumerate(func.patchings)`, because we have to delete func.patching attribute before try to re-apply patches otherwise new patch object will be added to patchings array of original function, so somehow we have to keep original array to store patch objects and delete reference from `func` and then should re-apply all patches.

wolever commented on parameterized/parameterized.py:
```diff
@@ -40,6 +40,31 @@ def make_method(func, instance, type):
+
+def delete_patches_if_need(func):
+    if hasattr(func, 'patchings'):
+        for _ in range(len(func.patchings)):
```
This can just be: `func.patchings[:] = []`

ukinau commented:
@wolever Thanks for looking at my PR and reviewing it.
I just changed my code as you reviewed and added unit test for the change.
As for `enumerate(func.patchings)` point we can't do this and I left a comment to explain the reason.
So could you check it out?

wolever commented on parameterized/parameterized.py:
```diff
@@ -40,6 +40,29 @@ def make_method(func, instance, type):
+        func = dummy_wrapper(func)
+        tmp_patchings = func.patchings
+        delattr(func, 'patchings')
+        for _, patch_obj in enumerate(tmp_patchings):
```
Is there a reason to use `enumerate` at all? (like, why not just `for patch_obj in tmp_patchings`)?

ukinau replied:
Nothing at all, sorry I will delete it.

wolever commented on parameterized/test.py:
```diff
@@ -101,6 +102,70 @@ def custom_naming_func(testcase_func, param_num, param):
+    expect([
+        "test_one_function_patch_decorator('foo1', 'umask', 'getpid')",
+        "test_one_function_patch_decorator('foo0', 'umask', 'getpid')",
+        "test_one_function_patch_decorator(42, 'umask', 'getpid')"])
```
Please put the closing `])` on a new line and add a trailing comma:

```
+        "test_one_function_patch_decorator(42, 'umask', 'getpid')",
+    ])
```

wolever commented on parameterized/test.py:
```diff
@@ -101,6 +102,70 @@ def custom_naming_func(testcase_func, param_num, param):
+        "test_multiple_function_patch_decorator"
+        "('foo0', 'bar0', 'umask', 'fdopen', 'getpid')",
+        "test_multiple_function_patch_decorator"
+        "('foo1', 'bar1', 'umask', 'fdopen', 'getpid')"])
```
Same as above re. `])`

wolever commented on parameterized/test.py:
```diff
@@ -101,6 +102,70 @@ def custom_naming_func(testcase_func, param_num, param):
+        "test_multiple_function_patch_decorator"
+        "('foo0', 'bar0', 'umask', 'fdopen')",
+        "test_multiple_function_patch_decorator"
+        "('foo1', 'bar1', 'umask', 'fdopen')"])
```
As above

wolever commented on parameterized/parameterized.py:
```diff
@@ -426,8 +449,16 @@ def parameterized_expand_wrapper(f, instance=None):
             for num, p in enumerate(paramters):
                 name = name_func(f, num, p)
-                frame_locals[name] = cls.param_as_standalone_func(p, f, name)
+                # If original function applied any patch, re-construct
```
Please change this to `# If the original function has patches applied by 'mock.patch', then re-construct…`
(ie, to make it explicit that these are patches from `mock`).

wolever commented:
Awesome!! I've left a couple small comments, and I'll merge + cut a release when you've got them done!

ukinau commented:
@wolever Thanks again. I changed as you review it. Could you check it?
